docs/language-reference/02-lexcial-structure.md

```diff
@@ -0,0 +1,114 @@
+Lexical Structure
+=================
+
+Source Units
+------------
+
+A _source unit_ comprises a sequence of zero or more _characters_ which for purposes of this document are defined as Unicode scalars (code points).
+Implementations *may* accept source units stored as files on disk, buffers in memory, or any appropriate implementation-specified means.
+
+Encoding
+--------
+
+When encoding is required, source units *should* be encoded using UTF-8.
+Implementations *may* support additional implemented-specified encodings.
+
+Whitespace
+----------
+
+_Horizontal whitespace_ consists of space (U+0020) and horizontal tab (U+0009).
+
+A _line break_ consists of a line feed (U+000A), carriage return (U+000D) or a carriage return followed by a line feed (U+000D, U+000A).
+Line breaks are used as line separators rather than terminators; it is not necessary for a source unit to end with a line break.
+
+Escaped Line Breaks
+-------------------
+
+An _escaped line break_ comprises a backslack (`\`, U+005C) follow immediately by a line break.
+
+Comments
+--------
+
+A _comment_ is either a line comment or a block comment.
+
+A _line comment_ comprises two forward slashes (`/`, U+002F) followed by zero or more characters that do not contain a line break.
+A line comment extends up to, but does not include, a subsequent line break or the end of the source unit.
+
+A _block comment_ begins with a forward slahs (`/`, U+002F) followed by an asterisk (`*`, U+0052). 
+A block comment is terminated by the next instance of an asterisk followed by a forward slash (`*/`).
+A block comment contains all characters between where it begins and where it terminates, including any line breaks.
+Block comments do not nest.
+It is an error if a block comment that begins in a source unit is not terminated in that source unit.
+
+Phases
+------
+
+Compilation of a source unit proceeds _as if_ the following steps are executed in order:
+
+1. Line numbering (for subsequent diagnostic messages) is noted based on the locations of line breaks
+
+2. Escaped line breaks are eliminated. No new characters are inserted to replace them. Any new escaped line breaks introduced by this step are not eliminated.
+
+3. All comments are replaced with a single space (U+0020)
+
+4. The source unit is _lexed_ into a sequence of tokens according the lexical grammar in this chapter
+
+5. The lexed sequence of tokens is _preprocessed_ to produce a new sequence of tokens (Chapter 3)
+
+6. Subsequent processing is performed on the preprocessed sequence of tokens
+
+Identifiers
+-----------
+
+An _identifier_ betgins with an upercase or lowercase ASCII leter (`A` through `Z`, `a` through `z`), or an underscore (`_`).
+After the first character, ASCII digits (`0` through `9`) may also be used in an identifier.
+
+The identifier consistent of a single underscore (`_`) is reserved by the language and must not be used by programs.
```

Comment:
One note (that I think will come up again later, based on my quick scan of your comments) is that there is an important difference between what the Slang front-end accepts and whatever constraints the different downstream compilers place.

The actual comment about a single underscore being reserved is currently just a goal, but I'd like to have documentation warn people that we might snag that identifier later.

No matter what we allow in identifiers (we could even allow non-ASCII code points if we felt like it...) we should be able to "scrub" them to something that a downstream compiler can accept. We already do this in cases where a "name hint" might include a `.` (which we turn into a `_`), and for GLSL we already deal with cases of identifiers with multiple underscores in a row (so that `__` turns into `_U_`, and `___` into `_U1_` or something like that, IIRC).

I don't think a reference for the Slang *language* should include those kinds of details, although it would be good if we had a per-back-end document like what you've added for CPU and CUDA to describe those kinds of details.